src/main/perf/DocGrouper.java

```diff
@@ -0,0 +1,438 @@
+            return null;
+        }
+
+        @Override
+        public int getNumOfDocsInGroup() {
+            return 0;
+        }
+
+        @Override
+        public int getRemainingNumGroups() {
+            return 0;
+        }
+    };
+
+    public static BytesRef[] group100;
+    public static BytesRef[] group100K;
+    public static BytesRef[] group10K;
+    public static BytesRef[] group1M;
+    public static BytesRef[] group2M;
+
+    DocGrouper(int numDocs, int targetNumGroups) {
+        this.numDocs = numDocs;
+        this.targetNumGroups = targetNumGroups;
+    }
+
+    public DocGroups getNextDocGroups() throws InterruptedException {
+        return outputQueue.take();
+    }
+
+    public static void initGroupIds(Random random) {
+        assert group100 == null;
+        group100 = randomStrings(100, random);
+        group10K = randomStrings(10000, random);
+        group100K = randomStrings(100000, random);
+        group1M = randomStrings(1000000, random);
+        group2M = randomStrings(2000000, random);
+    }
+
+    static int getTargetNumGroups(int numDocs) {
+        if (numDocs >= 5000000) {
+            return 1_000_000;
+        } else if (numDocs >= 500000) {
+            return 100_000;
+        } else {
+            return 10_000;
+        }
+    }
+
+    // returned array will not have dups
+    private static BytesRef[] randomStrings(int count, Random random) {
+        final BytesRef[] strings = new BytesRef[count];
+        HashSet<String> idSet = new HashSet<>(count);
+        int i = 0;
+        while(i < count) {
+            String s = randomRealisticUnicodeString(random);
+            while (s.equals("") == false && idSet.contains(s)) {
+                s = randomRealisticUnicodeString(random);
+            }
+            strings[i++] = new BytesRef(s);
+            idSet.add(s);
+        }
+
+        return strings;
+    }
+
+    public abstract void add(LineFileDocs.LineFileDoc lfd) throws InterruptedException;
+
+    /**
+     * A simple impl when we do not need grouping
+     */
+    static final class NoGroupImpl extends DocGrouper {
+
+        NoGroupImpl(int numDocs) {
+            super(numDocs, 0);
+        }
+
+        @Override
+        public void add(LineFileDocs.LineFileDoc lfd) throws InterruptedException {
+            if (lfd == LineFileDocs.END) {
+                outputQueue.put(END);
+            }
+            outputQueue.put(new DocGroups.SingleLFD(lfd));
+        }
+    }
+
+    static final class TextGrouper extends DocGrouper {
+
+        private int groupCounter;
+        private int docCounter;
+        private int nextNumDocs;
+        private LineFileDocs.LineFileDoc[] buffer;
+        private final BytesRef[] groupIds;
+        private final float docsPerGroupBlock;
+
+        TextGrouper(int numDocs) {
+            super(numDocs, getTargetNumGroups(numDocs));
+            assert group100 != null;
+            if (numDocs >= 5000000) {
+                groupIds = group1M;
```

Comment:
Hmm how come it's `group1M` here but `group2M` in the binary case?

Reply:
Oh because binary need some extra buffer... But anyway I'm going to remove binary LFD from grouping things so this will go away.